atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java

```diff
@@ -0,0 +1,113 @@
+import com.palantir.lock.CloseableRemoteLockService;
+import com.palantir.lock.LockClient;
+import com.palantir.lock.RemoteLockService;
+import com.palantir.timestamp.InMemoryTimestampService;
+
+public class SnapshotTransactionManagerTest {
+    private final CloseableRemoteLockService closeableRemoteLockService = mock(CloseableRemoteLockService.class);
+    private final Cleaner cleaner = mock(Cleaner.class);
+    private final KeyValueService keyValueService = mock(KeyValueService.class);
+
+    private final SnapshotTransactionManager snapshotTransactionManager = new SnapshotTransactionManager(
+            keyValueService,
+            new InMemoryTimestampService(),
+            LockClient.of("lock"),
+            closeableRemoteLockService,
+            null,
+            null,
+            null,
+            null,
+            cleaner);
+
+    @Test
+    public void closesKeyValueServiceOnClose() {
+        snapshotTransactionManager.close();
+        verify(keyValueService, times(1)).close();
+    }
+
+    @Test
+    public void closesCleanerOnClose() {
+        snapshotTransactionManager.close();
+        verify(cleaner, times(1)).close();
+    }
+
+    @Test
+    public void closesCloseableLockServiceOnClosingTransactionManager() throws IOException {
+        snapshotTransactionManager.close();
+        verify(closeableRemoteLockService, times(1)).close();
+    }
+
+    @Test
+    public void canCloseTransactionManagerWithNonCloseableLockService() {
+        SnapshotTransactionManager newTransactionManager = new SnapshotTransactionManager(
+                keyValueService,
+                new InMemoryTimestampService(),
+                LockClient.of("lock"),
+                mock(RemoteLockService.class), // not closeable
+                null,
+                null,
+                null,
+                null,
+                cleaner);
+        newTransactionManager.close(); // should not throw
+    }
+
+    @Test
+    public void cannotRegisterNullCallback() {
+        assertThatThrownBy(() -> snapshotTransactionManager.registerClosingCallback(null))
+                .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void invokesCallbackOnClose() {
+        Runnable callback = mock(Runnable.class);
+
+        snapshotTransactionManager.registerClosingCallback(callback);
+        snapshotTransactionManager.close();
+        verify(callback, times(1)).run();
+    }
+
+    @Test
+    public void invokesMultipleSuccessfulCallbacksInOrder() {
+        Runnable callback1 = mock(Runnable.class);
+        Runnable callback2 = mock(Runnable.class);
+        InOrder inOrder = inOrder(callback1, callback2);
+
+        snapshotTransactionManager.registerClosingCallback(callback1);
+        snapshotTransactionManager.registerClosingCallback(callback2);
+        snapshotTransactionManager.close();
+        inOrder.verify(callback1).run();
+        inOrder.verify(callback2).run();
```

Comment:
Hmm, yep now that you mention it close()s usually work in reverse order, so going in reverse order of registration seems to be more expected here